--- gc-7.3dev/configure	2013-08-18 01:34:06.000000000 +0000
+++ gc-7.3dev-patch/configure	2013-10-03 13:29:22.000000000 +0000
@@ -5124,7 +5124,7 @@
         THREADDLLIBS=-pthread
         INCLUDES="$INCLUDES -pthread"
         ;;
-     *-*-freebsd*)
+     *-*-freebsd*|*-*-dragonfly*)
         { $as_echo "$as_me:${as_lineno-$LINENO}: WARNING: \"FreeBSD does not yet fully support threads with Boehm GC.\"" >&5
 $as_echo "$as_me: WARNING: \"FreeBSD does not yet fully support threads with Boehm GC.\"" >&2;}
         $as_echo "#define GC_FREEBSD_THREADS 1" >>confdefs.h
*** gc-7.3dev/include/gc_config_macros.h	2013-08-18 03:32:54.000000000 +0200
--- gc-7.3dev-patch/include/gc_config_macros.h	2014-09-24 23:09:39.000000000 +0200
***************
*** 244,250 ****
    /* Only single-argument form of 'alloc_size' attribute is used.       */
  # if defined(__GNUC__) && (__GNUC__ > 4 \
          || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3 && !defined(__ICC)) \
!         || __clang_major__ > 3 \
          || (__clang_major__ == 3 && __clang_minor__ >= 2))
  #   define GC_ATTR_ALLOC_SIZE(argnum) __attribute__((__alloc_size__(argnum)))
  # else
--- 244,250 ----
    /* Only single-argument form of 'alloc_size' attribute is used.       */
  # if defined(__GNUC__) && (__GNUC__ > 4 \
          || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3 && !defined(__ICC)) \
!         || (__clang_major__ > 3 && __clang_major__ < 6 ) \
          || (__clang_major__ == 3 && __clang_minor__ >= 2))
  #   define GC_ATTR_ALLOC_SIZE(argnum) __attribute__((__alloc_size__(argnum)))
  # else
